packages/monaco/src/browser/monaco-semantic-highlighting-service.ts

```diff
@@ -0,0 +1,174 @@
+
+    protected readonly decorations = new Map<string, Set<string>>();
+    protected readonly toDisposeOnEditorClose = new Map<string, Disposable>();
+
+    async decorate(uri: URI, ranges: SemanticHighlightingRange[]): Promise<void> {
+        const editor = await this.editor(uri);
+        if (!editor) {
+            return;
+        }
+
+        const key = uri.toString();
+        if (!this.toDisposeOnEditorClose.has(key)) {
+            this.toDisposeOnEditorClose.set(key, new DisposableCollection(
+                editor.onDispose(() => this.deleteDecorations(key, editor)),
+                editor.onDocumentContentChanged(this.documentContentChanged.bind(this))
+            ));
+        }
+
+        const newDecorations = ranges.map(range => this.toDecoration(range));
+        const oldDecorations = this.oldDecorations(key, editor, ranges);
+        const newState = editor.deltaDecorations({
+            newDecorations,
+            oldDecorations
+        });
+
+        const decorationIds = this.decorationIds(key);
+        newState.forEach(id => decorationIds.add(id));
+
+        // Cache the new state.
+        this.decorations.set(key, decorationIds);
+    }
+
+    dispose(): void {
+        Array.from(this.toDisposeOnEditorClose.values()).forEach(disposable => disposable.dispose());
+    }
+
+    protected decorationIds(uri: string | URI): Set<string> {
+        return this.decorations.get(typeof uri === 'string' ? uri : uri.toString()) || new Set();
+    }
+
+    protected async editor(uri: string | URI): Promise<MonacoEditor | undefined> {
+        const editorWidget = await this.editorManager.getByUri(typeof uri === 'string' ? new URI(uri) : uri);
+        if (!!editorWidget && editorWidget.editor instanceof MonacoEditor) {
+            return editorWidget.editor;
+        }
+        return undefined;
+    }
+
+    protected async model(uri: string | URI): Promise<monaco.editor.ITextModel | undefined> {
+        const editor = await this.editor(uri);
+        if (editor) {
+            return editor.getControl().getModel();
+        }
+        return undefined;
+    }
+
+    /**
+     * Returns all the semantic highlighting decoration IDs that are affected by any of the range arguments.
+     */
+    protected oldDecorations(uri: string, editor: MonacoEditor, ranges: SemanticHighlightingRange[]): string[] {
+        const ids = this.decorationIds(uri);
+        const affectedLines = Array.from(new Set(ranges.map(r => [r.start.line, r.end.line]).reduce((prev, curr) => prev.concat(curr), [])));
+        return affectedLines
+            .map(line => editor.getLinesDecorations(line, line))
+            .reduce((prev, curr) => prev.concat(curr), [])
+            .map(decoration => decoration.id)
+            .filter(id => ids.has(id));
+    }
+
+    /**
+     * We do not get delta notification from the LS if lines were deleted and new semantic highlighting positions were introduced.
```

Comment:
I remember we talked about it, but I forget why that is. I think it should be the responsibility of the LS to send updates in that case, too.